src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -1959,6 +2023,146 @@ WorkerExtendedOpNode(MultiExtendedOp *originalOpNode,
 
 
 /*
+ * HasNonPartitionColumnDistinctAgg returns true if target entry or having qualifier
+ * has non-partition column reference in aggregate definition. Note that, it only
+ * checks aggs subfield of Aggref, it does not check FILTER or SORT clauses.
+ */
+static bool
+HasNonPartitionColumnDistinctAgg(List *targetEntryList, Node *havingQual,
+								 List *tableNodeList)
+{
+	List *targetVarList = pull_var_clause((Node *) targetEntryList,
+										  PVC_INCLUDE_AGGREGATES);
+	List *havingVarList = pull_var_clause((Node *) havingQual, PVC_INCLUDE_AGGREGATES);
+	List *aggregateCheckList = list_concat(targetVarList, havingVarList);
+
+	ListCell *aggregateCheckCell = NULL;
+	foreach(aggregateCheckCell, aggregateCheckList)
+	{
+		Node *targetNode = lfirst(aggregateCheckCell);
+		Aggref *targetAgg = NULL;
+		List *varList = NIL;
+		ListCell *varCell = NULL;
+		bool isPartitionColumn = false;
+
+		if (IsA(targetNode, Var))
+		{
+			continue;
+		}
+
+		Assert(IsA(targetNode, Aggref));
+		targetAgg = (Aggref *) targetNode;
+		if (targetAgg->aggdistinct == NIL)
+		{
+			continue;
+		}
+
+		varList = pull_var_clause_default((Node *) targetAgg->args);
+		foreach(varCell, varList)
+		{
+			Node *targetVar = (Node *) lfirst(varCell);
+			Assert(IsA(targetVar, Var));
+
+			isPartitionColumn =
+				HasPartitionColumn(targetVar, tableNodeList);
+
+			if (!isPartitionColumn)
+			{
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * ShouldPullDistinctColumn returns true if distinct aggregate should pull
+ * individual columns from worker to master and evaluate aggregate operation
+ * at master.
+ *
+ * Pull cases are:
+ * - repartition subqueries
+ * - query has count distinct on a non-partition column on at least one target
```

Comment:
Thinking about this a bit more, are not case-2 and case-3 same? Can we replace the second block with just;
```
	if (hasNonPartitionColumnDistinctAgg && !groupedByDisjointPartitionColumn)
	{
		return true;
	}
```

Reply:
not exactly the same
more like
```c
if (groupedByDisjointPartitionColumn)
{
    return false;
}

if (hasNonPartitionColumnDistinctAgg && !groupedByDisjointPartitionColumn)
{
    return true;
}

hasPartitionColumn = HasPartitionColumn(expression, tableNodeList);
if (!hasPartitionColumn)
{
    return true;
}
return false;

```

Reply:
What I am missing is that.
i. Do we suppose to call here for just count(distinct)?
ii. If yes, if `hasNonPartitionColumnDistinctAgg` is false, then `hasPartitionColumn` must be true?

Reply:
correct.  ```hasPartitionColumn``` and ``hasNonPartitionColumnDistinctAgg``` are not independent.

```!hasPartitionColumn``` implies ```hasNonPartitionColumnDistinctAgg```

We should be able remove ```hasPartitionColumn`` checks